dimod/bqm/src/shapeable.cc

```diff
@@ -0,0 +1,315 @@
+std::size_t num_interactions(const ShapeableBQM<Neighborhood, B> &bqm) {
+    std::size_t count = 0;
+    for (auto it = bqm.begin(); it != bqm.end(); ++it) {
+        count += (*it).first.size();
+    }
+    return count / 2;
+}
+
+
+template<class Neighborhood, class VarIndex, class Bias>
+Bias get_linear(const ShapeableBQM<Neighborhood, Bias> &bqm, VarIndex v) {
+    assert(v >= 0 && v < bqm.size());
+    return bqm[v].second;
+}
+
+
+template<class Neighborhood, class VarIndex, class Bias>
+std::pair<Bias, bool> get_quadratic(const ShapeableBQM<Neighborhood, Bias> &bqm,
+                                    VarIndex u, VarIndex v) {
+    assert(u >= 0 && u < bqm.size());
+    assert(v >= 0 && v < bqm.size());
+    assert(u != v);
+
+    // todo: we could search the shorter of the two neighborhoods
+
+    auto it = find_outvar(bqm[u].first, v);
+
+    if (it == bqm[u].first.end() || (*it).first != v)
+        return std::make_pair(0, false);
+
+    return std::make_pair((*it).second, true);
+}
+
+
+template<class Neighborhood, class VarIndex, class Bias>
+std::size_t degree(const ShapeableBQM<Neighborhood, Bias> &bqm, VarIndex v) {
+    return bqm[v].first.size();
+}
+
+
+// see note in header about why we need these two different functions
+
+template<class VarIndex, class Bias>
+std::pair<typename VectorNeighborhood<VarIndex, Bias>::iterator,
+          typename VectorNeighborhood<VarIndex, Bias>::iterator>
+neighborhood(AdjVectorBQM<VarIndex, Bias> &bqm, VarIndex v) {
+    assert(v >= 0 && v < bqm.size());
+    return std::make_pair(bqm[v].first.begin(), bqm[v].first.end());
+}
+
+template<typename VarIndex, typename Bias>
+std::pair<typename MapNeighborhood<VarIndex, Bias>::iterator,
+          typename MapNeighborhood<VarIndex, Bias>::iterator>
+neighborhood(AdjMapBQM<VarIndex, Bias> &bqm, VarIndex v) {
+    assert(v >= 0 && v < bqm.size());
+    return std::make_pair(bqm[v].first.begin(), bqm[v].first.end());
+}
+
+
+template<class Neighborhood, class VarIndex, class Bias>
+void set_linear(ShapeableBQM<Neighborhood, Bias> &bqm, VarIndex v, Bias b) {
+    assert(v >= 0 && v < bqm.size());
+    bqm[v].second = b;
+}
+
+
+template<typename VarIndex, typename Bias>
+void set_quadratic(AdjMapBQM<VarIndex, Bias> &bqm,
+                    VarIndex u, VarIndex v, Bias b) {
+    assert(u >= 0 && u < bqm.size());
+    assert(v >= 0 && v < bqm.size());
+    assert(u != v);
+
+    bqm[u].first[v] = b;
+    bqm[v].first[u] = b;
+}
+
+template<typename VarIndex, typename Bias>
+void set_quadratic(AdjVectorBQM<VarIndex, Bias> &bqm,
+                   VarIndex u, VarIndex v, Bias b) {
+    assert(u >= 0 && u < bqm.size());
+    assert(v >= 0 && v < bqm.size());
+    assert(u != v);
+
+    auto vit = find_outvar(bqm[u].first, v);
+    bool uv_exists = !(vit == bqm[u].first.end() || (*vit).first != v);
+    if (uv_exists) {
+        (*vit).second = b;
+    } else {
+        bqm[u].first.insert(vit, std::make_pair(v, b));
```

Comment:
```suggestion
        bqm[u].first.emplace(vit,v, b);
```
construct in place could be good